### DISCUSSION TITLE
Full MPS simulation of quantum machine learning model

### DISCUSSION THREAD

**Discussion by lolazon22** [tencent-quantum-lab/tensorcircuit#84] [Q&A]
Hello,
I plan to implement a quantum machine learning model for some classification problems. The circuit is very shallow, but the number of qubits is relatively large, say >30. I tried to follow the beginner tutorial, and I implemented the circuit and optimization procedure:

```
n = 30
layers = 1

def loss(params, nqubits):
    loss = 0.0
    for k in range(len(ising_groundstates)):
        c = tc.Circuit(nqubits, mps_inputs=ising_groundstates[k])
        i = 0
        for l in range(layers):
            for q in range(nqubits):
                c.ry(q, theta=params[i])
                i+=1
            for q in range(0, nqubits - 1, 2):
                c.cz(q, q+1)
            for q in range(nqubits):
                c.ry(q, theta=params[i])
                i+=1
            for q in range(1, nqubits - 2, 2):
                c.cz(q, q+1)
            c.cz(0, nqubits-1)        
        c.ry(0, theta=params[i])
            
        for m in range(nqubits):
            loss += c.expectation([tc.gates.z(), [m]])
    return K.real(loss)

ising_groundstates = [] # here, one should generate a list of TFIM ground states for different lambdas

vgf = K.jit(K.value_and_grad(loss), static_argnums=1)
params = tf.Variable(tf.random.normal(stddev=0.1, shape=[2*layers*n+1]))
opt = tc.backend.optimizer(tf.keras.optimizers.Adam(learning_rate=0.1))
for j in range(5000):
    loss, gr = vgf(params, n)
    param = opt.update(gr, params)
    if j % 50 == 0:
        print("loss", loss.numpy())
```

I am facing two main issues with the code. 
1. First, I do not manage to run the full simulation as an MPS. Indeed, since the circuit is very shallow, the MPS simulation would run efficiently. Otherwise, I wouldn't be able to run circuits with >30 qubits.
2. The input states of the variational circuit are ground states of the 1D TFIM, which are efficiently prepared as MPS since the entanglement is low. I would like to know whether there exists the possibility to create them as an MPS with TensorCircuit, or otherwise, I have to use other packages and import the MPS state (if that is the case, do you have any hint?).

I appreciate your help!

**Comment by refraction-ray** [comment]
1. As long as the `mps_inputs` is in the format of `quvector` (MPS in tensorcircuit), I believe you can simulate system with more than 30 qubits since `nlayers=1`. Besides, if the `nlayers` is larger, one can also use `tc.MPSCircuit` which provides a MPS TEBD simulator and can simulate more qubits with some approximation truncation trade off. For `Circuit`, the input argument for the constructor is `mps_inputs` which accepts `quvector`; and for `MPSCircuit`, the input argument can be `tensors` which accepts a list of tensor for the MPS or `wavefunction` which accepts `quvector`.

2. There are two steps to prepare the MPS inputs. 1) obtain the groud state MPS via DMRG in some packages, here I suggest quimb: see https://quimb.readthedocs.io/en/latest/examples/ex_dmrg_periodic.html for an example of DMRG. 2) We can transform the MPS class in quimb to `quvector` in tensorcircuit via function provided by tensorcircuit (`tc.quantum.quimb2qop`), see https://github.com/tencent-quantum-lab/tensorcircuit/blob/master/tests/test_quantum.py#L391-L413 for an example.

Besides, currently in the code, the loss is not correct for supervised learning where instead of expectation on z (ypred), you need further to construct the MSE or crossentropy as the final loss to do gradient descent to train the phase classifier for TFIM.

**Reply by lolazon22** [reply]
> 1. As long as the `mps_inputs` is in the format of `quvector` (MPS in tensorcircuit), I believe you can simulate system with more than 30 qubits since `nlayers=1`. Besides, if the `nlayers` is larger, one can also use `tc.MPSCircuit` which provides a MPS TEBD simulator and can simulate more qubits with some approximation truncation trade off. For `Circuit`, the input argument for the constructor is `mps_inputs` which accepts `quvector`; and for `MPSCircuit`, the input argument can be `tensors` which accepts a list of tensor for the MPS or `wavefunction` which accepts `quvector`.
> 
>     2. There are two steps to prepare the MPS inputs. 1) obtain the groud state MPS via DMRG in some packages, here I suggest quimb: see https://quimb.readthedocs.io/en/latest/examples/ex_dmrg_periodic.html for an example of DMRG. 2) We can transform the MPS class in quimb to `quvector` in tensorcircuit via function provided by tensorcircuit (`tc.quantum.quimb2qop`), see https://github.com/tencent-quantum-lab/tensorcircuit/blob/master/tests/test_quantum.py#L391-L413 for an example.
> 
> 
> Besides, currently in the code, the loss is not correct for supervised learning where instead of expectation on z (ypred), you need further to construct the MSE or crossentropy as the final loss to do gradient descent to train the phase classifier for TFIM.

Thank you, @refraction-ray. I will follow your recommendations. I will try to implement everything in the following days.

**Reply by lolazon22** [reply]
Hello @refraction-ray. I applied all your recommendations, except for the last part, where I do have to transform the MPS class from quimb to `quvector`. See an example code below:

```
import tensorcircuit as tc
import tensorflow as tf
from quimb import *
from quimb.tensor import *

# Create Hamiltonian with 32 spins
H = MPO_ham_heis(32, cyclic=False)
dmrg = DMRG2(H)
# Solve DMRG
dmrg.solve(max_sweeps=4, verbosity=1, cutoffs=1e-6)
# Get groundstate
gs = dmrg.state
# Transform to quvector
gs_tc = tc.quantum.quimb2qop(gs)
```
I get the following error: `raise ValueError("Cannot connect edges of unequal dimension. "
ValueError: Cannot connect edges of unequal dimension. Dimension of edge '__unnamed_edge__': 2, Dimension of edge '__unnamed_edge__': 4.`

Any help would be appreciated. Thanks.

**Reply by refraction-ray** [reply]
is the error raised in ``quimb2qop`` or later usage? could you provide a minimal but complete demo to reproduce the problem?

**Reply by lolazon22** [reply]
@refraction-ray the error is raised by `tensorcircuit`. This code works, i.e. producing the ground state with `quimb`:

```
import tensorcircuit as tc
import tensorflow as tf
from quimb import *
from quimb.tensor import *

# Create Hamiltonian with 32 spins
H = MPO_ham_heis(32, cyclic=False)
dmrg = DMRG2(H)
# Solve DMRG
dmrg.solve(max_sweeps=4, verbosity=1, cutoffs=1e-6)
# Get groundstate
gs = dmrg.state
```

Then, I try to apply `quimb2qop` to transform to `quvector`, and I get the following error:

```
# Transform to quvector
gs_tc = tc.quantum.quimb2qop(gs)
```

```
File "/TensorCircuit_MPS_try.py", line 19, in <module>
    gs_tc = tc.quantum.quimb2qop(gs)
  File "/tensorcircuit/quantum.py", line 1096, in quimb2qop
    connect(mpo[i][1], mpo[i + 1][0])
  File "/tensornetwork/network_components.py", line 1949, in connect
    raise ValueError("Cannot connect edges of unequal dimension. "
ValueError: Cannot connect edges of unequal dimension. Dimension of edge '__unnamed_edge__': 2, Dimension of edge '__unnamed_edge__': 4.
```

**Reply by refraction-ray** [reply]
thanks for your detailed demo, I have identified the bug: ``quimb2qop`` has some drawback: when the input quimb tensors are not in [bond, bond, outer] dimension, will fix soon.

**Reply by refraction-ray** [reply]
the latest commit on Github should fix the bug and work as expected, please have a try

**Reply by lolazon22** [reply]
It works like a charm! If you do not mind, I will leave the discussion open for a few days in case I encounter any further issues or I still have questions. Thanks @refraction-ray

**Reply by refraction-ray** [reply]
sure, feel free to keep it open and ask me for further questions anytime

**Reply by lolazon22** [reply]
Hello @refraction-ray, I am here again just a formulate a quick question. For the simulations with 30+ qubits, I plan to implement a circuit similar to this one here (Figure 1b [here](https://arxiv.org/pdf/1810.03787.pdf)), where you measure qubits now and then and act with gates depending on the measurement. I was wondering whether this is possible to implement with TensorCircuit using the MPS backend, and if it helps with the simulation for large number of qubits since, effectively, after the measurement, you will reduce the size of the circuit. I tried to look for information in the documentation but I did not see anything. Thanks in advance!

**Reply by refraction-ray** [reply]
@lolazon22 , please check whether https://github.com/tencent-quantum-lab/tensorcircuit/issues/52#issuecomment-1193272784 helps. Note that in MPSCircuit case, only single and two - qubit gates are supported. But I guess it is enough for the vanilla QCNN implementation.

**Reply by lolazon22** [reply]
@refraction-ray Thanks for the reply. Indeed, what you comment in there (deferred measurement) is actually what I was implementing for the circuits with fewer qubits. However, for the +30 qubit case, I cannot run the circuit anymore because I do not have enough memory (I guess I could use the approximation truncation worst-case scenario). I believe, in this case, actual measurements (instead of deferred measurement) would help to reduce the effective number of qubits in the circuit and allow for the exact simulation with MPS.

Anyhow, I'll take a look into the approximation truncation. Thanks!

**Reply by lolazon22** [reply]
@refraction-ray I am mostly done with all my numerics, but a new error is popping up when using MPSCircuit for large number of qubits. Let me put an example that works:

```python
import numpy as np
import tensorcircuit as tc
import tensorflow as tf
from quimb import *
from quimb.tensor import *

nqubits=32

# Create Hamiltonian
H = MPO_ham_heis(nqubits, cyclic=False)
dmrg = DMRG2(H)
# Solve DMRG
dmrg.solve(max_sweeps=4, verbosity=1, cutoffs=1e-6)
# Get groundstate
gs = dmrg.state
# Transform to quvector
gs_tc = tc.quantum.quimb2qop(gs)

# Create circuit
c = tc.Circuit(nqubits, mps_inputs=gs_tc)
```

The previous code actually works well. However, as I mentioned in the previous discussion, the circuit I am going to implement is not very shallow, and I want to implement an approximation in the truncation (otherwise, the bond dimension will grow too much) by using `tc.MPSCircuit` instead of `tc.Circuit`. Thus, I am using the following code:

```python
import numpy as np
import tensorcircuit as tc
import tensorflow as tf
from quimb import *
from quimb.tensor import *

nqubits=32

# Create Hamiltonian
H = MPO_ham_heis(nqubits, cyclic=False)
dmrg = DMRG2(H)
# Solve DMRG
dmrg.solve(max_sweeps=4, verbosity=1, cutoffs=1e-6)
# Get groundstate
gs = dmrg.state
# Transform to quvector
gs_tc = tc.quantum.quimb2qop(gs)

c = tc.MPSCircuit(nqubits, wavefunction=gs_tc)
c.set_split_rules({"max_singular_values": 20})
```

But the following error appears:
```

Traceback (most recent call last):
  File "TensorCircuit_MPS.py", line 30, in <module>
    c = tc.MPSCircuit(nqubits, wavefunction=gs_tc)
  File "/.local/lib/python3.10/site-packages/tensorcircuit/mpscircuit.py", line 115, in __init__
    wavefunction = wavefunction.eval()
  File "/.local/lib/python3.10/site-packages/tensorcircuit/quantum.py", line 705, in eval
    self.contract(final_edge_order)
  File "/.local/lib/python3.10/site-packages/tensorcircuit/quantum.py", line 678, in contract
    [contractor(self.nodes, output_edge_order=final_edge_order)]
  File "/.local/lib/python3.10/site-packages/tensorcircuit/cons.py", line 654, in custom
    return _base(
  File "/.local/lib/python3.10/site-packages/tensorcircuit/cons.py", line 614, in _base
    new_node = tn.contract_between(nodes[a], nodes[b], allow_outer_product=True)
  File "/.local/lib/python3.10/site-packages/tensornetwork/network_components.py", line 2085, in contract_between
    new_tensor = backend.tensordot(node1.tensor, node2.tensor, [axes1, axes2])
  File "/.local/lib/python3.10/site-packages/tensornetwork/backends/numpy/numpy_backend.py", line 53, in tensordot
    return np.tensordot(a, b, axes)
  File "<__array_function__ internals>", line 180, in tensordot
  File "/.local/lib/python3.10/site-packages/numpy/core/numeric.py", line 1138, in tensordot
    res = dot(at, bt)
  File "<__array_function__ internals>", line 180, in dot
numpy.core._exceptions._ArrayMemoryError: Unable to allocate 32.0 GiB for an array with shape (8192, 524288) and data type float64
```
It feels like TensorCircuit forgets about the fact that `gs_tc` is an MPS and tries to momentarily save the entire state vector in memory for some reason. Any help would be appreciated! Thank you very much.

**Reply by refraction-ray** [reply]
@lolazon22 ， yes for the input part, MPSCircuit currently has no consistent interface with Circuit, while `mps_inputs` can takes a MPS as wavefunction in Circuit case, in `MPSCircuit` case, we have to use `tensor` instead of `wavefunction` for the MPS input, which is just the list of tensors in the MPS

**Reply by lolazon22** [reply]
@refraction-ray thanks for the quick reply! I believe then I should not use `gs_tc = tc.quantum.quimb2qop(gs)` and directly use the tensors right? Do you know which should be the input in the previous example? I tried with `gs` but error pops up as well. I am a bit lost here.

**Reply by refraction-ray** [reply]
@lolazon22, ``[n.tensor for n in gs_tc.nodes]`` but please double check that the tensor list is in the correct order

**Reply by lolazon22** [reply]
Thanks @refraction-ray. Unfortunately, this error pops up:

```python
import numpy as np
import tensorcircuit as tc
import tensorflow as tf
from quimb import *
from quimb.tensor import *

nqubits=32

# Create Hamiltonian
H = MPO_ham_heis(nqubits, cyclic=False)
dmrg = DMRG2(H)
# Solve DMRG
dmrg.solve(max_sweeps=4, verbosity=1, cutoffs=1e-6)
# Get groundstate
gs = dmrg.state
# Transform to quvector
gs_tc = tc.quantum.quimb2qop(gs)
gs_tc = [n.tensor for n in gs_tc.nodes]

c = tc.MPSCircuit(nqubits, tensors=gs_tc)
c.set_split_rules({"max_singular_values": 20})
```
```
Traceback (most recent call last):
  File "TensorCircuit_MPS.py", line 31, in <module>
    c = tc.MPSCircuit(nqubits, tensors=gs_tc)
  File "/.local/lib/python3.10/site-packages/tensorcircuit/mpscircuit.py", line 127, in __init__
    self._mps = FiniteMPS(tensors, canonicalize=True, center_position=0)
  File "/.local/lib/python3.10/site-packages/tensornetwork/matrixproductstates/finite_mps.py", line 84, in __init__
    self.position(len(self) - 1)
  File "/.local/lib/python3.10/site-packages/tensornetwork/matrixproductstates/base_mps.py", line 190, in position
    self.tensors[n + 1] = ncon.ncon([rest, self.tensors[n + 1]],
  File "/.local/lib/python3.10/site-packages/tensornetwork/ncon_interface.py", line 634, in ncon
    _check_network(network_structure, [t.shape for t in _tensors], con_order,
  File "/.local/lib/python3.10/site-packages/tensornetwork/ncon_interface.py", line 237, in _check_network
    raise ValueError(
ValueError: tensor dimensions for labels [1] are mismatching
```
Any possible idea? I guess this is related to what you mentioned about being in the correct order. I tried to fix it, but I have not succeeded.

**Reply by refraction-ray** [reply]
The input argument for mpscircuit is not wacefunction but tensor

**Reply by lolazon22** [reply]
I actually copied/pasted the previous comment that's why it shows `wavefunction` instead of `tensors`, my bad. I edited it, so it shows `tensors`. The error indeed appears when using `tensors`. Any possible suggestion? Thanks! @refraction-ray

**Reply by refraction-ray** [reply]
> I actually copied/pasted the previous comment that's why it shows `wavefunction` instead of `tensors`, my bad. I edited it, so it shows `tensors`. The error indeed appears when using `tensors`. Any possible suggestion? Thanks! @refraction-ray

pls try the following which works for me

```python
tensors = [t.data for t in gs.tensors]
tensors[0] = tensors[0].reshape([1]+list(tensors[0].shape))
tensors[-1] = tensors[-1].reshape(list(tensors[-1].shape)+[1])

c = tc.MPSCircuit(32, tensors=tensors)
```

**Reply by lolazon22** [reply]
Thanks @refraction-ray, it solves the issue partially. As soon as you implement a measurement to compute the expectation value of some MPO it crashes again due to memory issues (the last step I need!). It kills the process. Any possible workaround? I believe this increase in memory should not happen, right? Find below an example so you can check it as well.

```python
import numpy as np
import tensorcircuit as tc
import tensorflow as tf
from quimb import *
from quimb.tensor import *

nqubits=32

# Create Hamiltonian
H = MPO_ham_heis(nqubits, cyclic=False)
dmrg = DMRG2(H)
# Solve DMRG
dmrg.solve(max_sweeps=4, verbosity=1, cutoffs=1e-6)
# Get groundstate
gs = dmrg.state

tensors = [t.data for t in gs.tensors]
tensors[0] = tensors[0].reshape([1]+list(tensors[0].shape))
tensors[-1] = tensors[-1].reshape(list(tensors[-1].shape)+[1])

c = tc.MPSCircuit(nqubits, tensors=tensors)
c.set_split_rules({"max_singular_values": 20})

# Expectation value sigma_z ... sigma_z
z0, z1, z2, z3, z4, z5, z6, z7, z8, z9, z10, z11, z12, z13, z14, z15, z16, z17, z18, z19, z20, z21, z22, z23, z24, z25, z26, z27, z28, z29, z30, z31 = tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z(), tc.gates.z()

mpo = tc.quantum.QuOperator([z0[0], z1[0], z2[0], z3[0], z4[0], z5[0], z6[0], z7[0], z8[0], z9[0], z10[0], z11[0], z12[0], z13[0], z14[0], z15[0], z16[0], z17[0], z18[0], z19[0], z20[0], z21[0], z22[0], z23[0], z24[0], z25[0], z26[0], z27[0], z28[0], z29[0], z30[0], z31[0]], [z0[1], z1[1], z2[1], z3[1], z4[1], z5[1], z6[1], z7[1], z8[1], z9[1], z10[1], z11[1], z12[1], z13[1], z14[1], z15[1], z16[1], z17[1], z18[1], z19[1], z20[1], z21[1], z22[1], z23[1], z24[1], z25[1], z26[1], z27[1], z28[1], z29[1], z30[1], z31[1]])
   
print(tc.templates.measurements.mpo_expectation(c, mpo))
```

**Reply by refraction-ray** [reply]
``c.expectation_ps(z=[i for i in range(32)])`` try this to measure instead

**Reply by lolazon22** [reply]
It works, thanks @refraction-ray !

**Reply by lolazon22** [reply]
Hello @refraction-ray. I was about to open an issue to report a possible bug, but I believe it might be too specific to my implementation. As you know, I am currently using `MPSCircuit` to generate circuits in an optimization loop to minimize a cost function, and `quimb` to generate the initial states. It happens that, eventually, after running for several hours the optimization, the following error appears and kills the program, so it is impossible to complete the optimization. I will not attach the code since it may be too large, but if needed I can do so. The error falls back to `numpy`, but I believe it has to do with the implementation of `MPSCircuit`.

```
File "/code.py", line 63, in pooling_layer
    c.cr(q1, q2, theta=param[0], alpha=param[1], phi=param[2])
  File "/.local/lib/python3.9/site-packages/tensorcircuit/abstractcircuit.py", line 122, in apply
    self.apply_general_gate(
  File "/.local/lib/python3.9/site-packages/tensorcircuit/mpscircuit.py", line 581, in apply_general_gate
    self.apply_double_gate(gate, *index, split=split)  # type: ignore
  File "/.local/lib/python3.9/site-packages/tensorcircuit/mpscircuit.py", line 305, in apply_double_gate
    self.apply_adjacent_double_gate(
  File "/.local/lib/python3.9/site-packages/tensorcircuit/mpscircuit.py", line 242, in apply_adjacent_double_gate
    err = self._mps.apply_two_site_gate(
  File "/.local/lib/python3.9/site-packages/tensorcircuit/mps_base.py", line 132, in apply_two_site_gate
    U, S, V, tw = self.backend.svd(
  File "/.local/lib/python3.9/site-packages/tensornetwork/backends/numpy/numpy_backend.py", line 622, in svd
    return decompositions.svd(np,
  File "/.local/lib/python3.9/site-packages/tensornetwork/backends/numpy/decompositions.py", line 36, in svd
    u, s, vh = np.linalg.svd(tensor, full_matrices=False)
  File "<__array_function__ internals>", line 180, in svd
  File "/net/opt/python/stretch/3.9.7/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 1648, in svd
    u, s, vh = gufunc(a, signature=signature, extobj=extobj)
  File "/net/opt/python/stretch/3.9.7/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 97, in _raise_linalgerror_svd_nonconvergence
    raise LinAlgError("SVD did not converge")
numpy.linalg.LinAlgError: SVD did not converge
```

**Reply by refraction-ray** [reply]
According to [this](https://stackoverflow.com/questions/21827594/raise-linalgerrorsvd-did-not-converge-linalgerror-svd-did-not-converge-in-m), this may be due to there is nan in the data, maybe you need to reduce the demo in terms of the code size and reproducing time, so that the bug (if any) can be identified and fixed, otherwise, nan could be somehow common if singular matrix occurs during optimization

**Reply by refraction-ray** [reply]
Another possibility is [this](https://stackoverflow.com/a/57245768/9062180) due to memory leakage somewhere. The workaround is, if the crash time is rather fixed, we can first store the weights after some iterations and kill the program. And then open a new program loading the weights. Identifying a place that leak the memory is very challenging otherwise

**Reply by refraction-ray** [reply]
or [this](https://stackoverflow.com/a/37714528/9062180) I think you need to be patient enough to print intermediate variable and wait for the crash to identify what is the reason first
